crates/sui-network/src/state_sync/mod.rs

```diff
@@ -594,22 +646,35 @@ async fn query_peers_for_their_latest_checkpoint(
     let futs = peer_heights
         .read()
         .unwrap()
-        .heights
-        .keys()
+        .peers_on_same_chain()
         // Filter out any peers who we aren't connected with
-        .flat_map(|peer_id| network.peer(*peer_id))
-        .map(|peer| {
+        .flat_map(|(peer_id, info)| network.peer(*peer_id).map(|peer| (peer, *info)))
+        .map(|(peer, mut info)| {
             let peer_id = peer.peer_id();
             let mut client = StateSyncClient::new(peer);
 
-            let request =
-                Request::new(GetCheckpointSummaryRequest::Latest).with_timeout(DEFAULT_TIMEOUT);
             async move {
+                let request =
+                    Request::new(GetCheckpointSummaryRequest::Latest).with_timeout(DEFAULT_TIMEOUT);
                 let response = client
                     .get_checkpoint_summary(request)
                     .await
                     .map(Response::into_inner);
-                update_peer_height(peer_heights, peer_id, &response);
+                match &response {
+                    Ok(Some(checkpoint)) => {
+                        info.height = std::cmp::max(checkpoint.sequence_number(), info.height);
```

Comment:
is this thread safe? it seems like you make a copy of PeerStateSyncInfo under a read lock, then later update based on that copy (and not the current value) after acquiring the write lock? could this overwrite newer values with older values?

Reply:
So yes this is "thread safe" as in there's no undefined behavior happening. But you're correct that there could be some odd races. I've adjusted the PR to eliminate these so it should be clearer what should be happening.

Reply:
Since pedantry is a favorite engineering pastime 🤓🤓🤓...I was using "thread safe" in the sense of "free of race conditions," which at least from Wikipedia looks like a valid use of the term to me.